tests/riotboot/tests/01-run.py

```diff
@@ -54,27 +54,45 @@ def assert_check_slot(child, slotnum, version):
     child.expect_exact('>')
 
 
-def testfunc(child):
-    # Ask for current slot, should be 0 or 1
-    child.expect_exact('>')
-    child.sendline("curslotnr")
-    child.expect(r"Current slot=([0-1])")
-    slotnum = int(child.match.group(1))
+class riotbootDevice():
+    def __init__(self):
+        self.ret = 0
+        self.ret |= run(self._get_current)
 
-    # Ask for current APP_VER
-    child.expect_exact('>')
-    child.sendline("curslothdr")
-    child.expect(r"Image Version: (?P<app_ver>0x[0-9a-fA-F]{8})\r\n")
-    current_app_ver = int(child.match.group("app_ver"), 16)
+    def _get_current(self, child):
+        # Ask for current slot, should be 0 or 1
+        child.expect_exact('>')
+        child.sendline("curslotnr")
+        child.expect(r"Current slot=([0-1])")
+        self.slot_num = int(child.match.group(1))
 
-    # Flash to both slots and verify basic functions
-    for version in [current_app_ver + 1, current_app_ver + 2]:
-        slotnum = slotnum ^ 1
-        flash_slot(slotnum, version)
-        assert_check_slot(child, slotnum, version)
+        # Ask for current APP_VER
+        child.expect_exact('>')
+        child.sendline("curslothdr")
+        child.expect(r"Image Version: (?P<app_ver>0x[0-9a-fA-F]{8})\r\n")
+        self.app_ver = int(child.match.group("app_ver"), 16)
 
-    print("TEST PASSED")
+    def test_current(self, child):
+        # Verify that running slot_num and app_ver match the stored ones
+        assert_check_slot(child, self.slot_num, self.app_ver)
+        print("TEST PASSED")
 
 
 if __name__ == "__main__":
-    sys.exit(run(testfunc))
+    # This test requires flashing slot binaries. Some flashers like
+    # cc2538-bsl need to attach to the serial terminal in order to be
+    # able to flash. Since `run(testfunc)` spawns a terminal a different
+    # `run(testfunc)` is called after every flash in order to detach and
+    # re-attach the terminal between flashes
+    dev = riotbootDevice()
+    # flash to both slots and verify
+    for version in [dev.app_ver + 1, dev.app_ver + 2]:
+        # update the stored app_ver and slot_num
+        dev.slot_num = dev.slot_num ^ 1
+        dev.app_ver = version
+        # flash
+        flash_slot(dev.slot_num, dev.app_ver)
+        # verify its running from the correct slot with the right version
+        dev.ret |= run(dev.test_current)
+
+    sys.exit(dev.ret)
```

Comment:
```suggestion
    sys.exit(testfunc())
```

And testfunc would be like (I removed the comments on purpose so the snippet is smaller, but I hope you'll get the idea):

```python
def testfunc():
    dev = RiotbootDevice()
    ret = run(dev->get_current_slot_and_app_ver)
    if ret != 0:
        return ret

    for version in [dev.app_ver + 1, dev.app_ver + 2]:
        dev.slot_num = dev.slot_num ^ 1
        dev.app_ver = version
        dev.flash_current_slot()
        ret = run(dev.check_current_slot_and_app_ver)
        if ret != 0:
            return ret
    return 0
```

Doing like this also has the benefit to fail fast if something goes wrong. And `flash_slot` (renamed as `flash_current_slot`) would be a method of `RiotbootDevice` and I would rename `test_current` as `check_current_slot_and_app_ver` which is a more explicit name.